File: android/app/src/main/java/com/magicforms/mobile/data/OfflineSubmitQueue.kt
package com.magicforms.mobile.data

import android.content.Context
import com.magicforms.mobile.data.api.FormSubmitApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

@Serializable
data class PendingFormSubmit(
    val id: String,
    val kind: String,
    val formId: Int? = null,
    val relatedAccessToken: String? = null,
    val textFields: Map<String, String> = emptyMap(),
    val checkboxFields: Map<String, Boolean> = emptyMap(),
    val checklistFields: Map<String, List<String>> = emptyMap(),
)

class OfflineSubmitQueue(context: Context) {
    private val prefs = context.getSharedPreferences("offline_submit_queue", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }
    private val api = FormSubmitApi()

    fun pendingCount(): Int = load().size

    fun enqueue(item: PendingFormSubmit) {
        val items = load().toMutableList()
        items.add(item)
        save(items)
    }

    fun flush(token: String): Int {
        val items = load().toMutableList()
        if (items.isEmpty()) return 0
        var synced = 0
        val remaining = mutableListOf<PendingFormSubmit>()
        for (item in items) {
            try {
                when {
                    item.kind == "related" && item.relatedAccessToken != null -> {
                        api.submitRelated(
                            token,
                            item.relatedAccessToken,
                            item.textFields,
                            item.checkboxFields,
                            item.checklistFields,
                            emptyMap(),
                        )
                    }
                    item.formId != null -> {
                        api.submit(
                            token,
                            item.formId,
                            item.textFields,
                            item.checkboxFields,
                            item.checklistFields,
                            emptyMap(),
                        )
                    }
                    else -> {
                        remaining.add(item)
                        continue
                    }
                }
                synced++
            } catch (_: Exception) {
                remaining.add(item)
            }
        }
        save(remaining)
        return synced
    }

    fun newItem(
        kind: String,
        formId: Int?,
        relatedAccessToken: String?,
        textFields: Map<String, String>,
        checkboxFields: Map<String, Boolean>,
        checklistFields: Map<String, List<String>>,
    ) = PendingFormSubmit(
        id = UUID.randomUUID().toString(),
        kind = kind,
        formId = formId,
        relatedAccessToken = relatedAccessToken,
        textFields = textFields,
        checkboxFields = checkboxFields,
        checklistFields = checklistFields,
    )

    private fun load(): List<PendingFormSubmit> {
        val raw = prefs.getString("items", null) ?: return emptyList()
        return runCatching { json.decodeFromString<List<PendingFormSubmit>>(raw) }.getOrElse { emptyList() }
    }

    private fun save(items: List<PendingFormSubmit>) {
        prefs.edit().putString("items", json.encodeToString(items)).apply()
    }
}

File: ios/MagicFormsMobile/Services/OfflineSubmitQueue.swift
import Foundation

struct PendingFormSubmit: Codable, Identifiable {
    let id: String
    let kind: String
    let formId: Int?
    let relatedAccessToken: String?
    let textFields: [String: String]
    let checkboxFields: [String: Bool]
    let checklistFields: [String: [String]]
    let createdAt: Date
}

@MainActor
final class OfflineSubmitQueue: ObservableObject {
    static let shared = OfflineSubmitQueue()

    @Published private(set) var pendingCount = 0

    private let storageKey = "offline_form_submit_queue"

    private init() {
        pendingCount = load().count
    }

    func enqueue(_ item: PendingFormSubmit) {
        var items = load()
        items.append(item)
        save(items)
        pendingCount = items.count
    }

    func flush(authToken: String) async -> Int {
        let items = load()
        guard !items.isEmpty else { return 0 }
        var synced = 0
        var remaining: [PendingFormSubmit] = []

        for item in items {
            do {
                let files: [String: (data: Data, fileName: String, mimeType: String)] = [:]
                if item.kind == "related", let token = item.relatedAccessToken {
                    _ = try await FormSubmitAPI.submitRelated(
                        token: authToken,
                        accessToken: token,
                        textFields: item.textFields,
                        checkboxFields: item.checkboxFields,
                        checklistFields: item.checklistFields,
                        files: files
                    )
                } else if let formId = item.formId {
                    _ = try await FormSubmitAPI.submit(
                        token: authToken,
                        formId: formId,
                        textFields: item.textFields,
                        checkboxFields: item.checkboxFields,
                        checklistFields: item.checklistFields,
                        files: files
                    )
                } else {
                    remaining.append(item)
                    continue
                }
                synced += 1
            } catch {
                remaining.append(item)
            }
        }
        save(remaining)
        pendingCount = remaining.count
        return synced
    }

    private func load() -> [PendingFormSubmit] {
        guard let data = UserDefaults.standard.data(forKey: storageKey) else { return [] }
        return (try? JSONDecoder().decode([PendingFormSubmit].self, from: data)) ?? []
    }

    private func save(_ items: [PendingFormSubmit]) {
        guard let data = try? JSONEncoder().encode(items) else { return }
        UserDefaults.standard.set(data, forKey: storageKey)
    }
}
